File: elixir/lib/algorithms/insertion_sort.ex
defmodule Algorithms.InsertionSort do
  @moduledoc """
  Insertion sort algorithm.
  Time: O(n²)
  Space: O(n) - creates new list
  """

  @doc "Sort a list using insertion sort"
  @spec sort(list()) :: list()
  def sort(list) do
    Enum.reduce(list, [], &insert_sorted/2)
  end

  @doc "Sort with custom comparison function"
  @spec sort_by(list(), (any(), any() -> boolean())) :: list()
  def sort_by(list, compare) do
    Enum.reduce(list, [], fn x, acc -> insert_sorted_by(x, acc, compare) end)
  end

  defp insert_sorted(x, []), do: [x]

  defp insert_sorted(x, [head | tail] = sorted) do
    if x <= head do
      [x | sorted]
    else
      [head | insert_sorted(x, tail)]
    end
  end

  defp insert_sorted_by(x, [], _compare), do: [x]

  defp insert_sorted_by(x, [head | tail] = sorted, compare) do
    if compare.(x, head) do
      [x | sorted]
    else
      [head | insert_sorted_by(x, tail, compare)]
    end
  end
end

File: haskell/src/Algorithms/InsertionSort.hs
{-|
Module      : Algorithms.InsertionSort
Description : Insertion sort algorithm
Stability   : stable

A purely functional insertion sort implementation.
Time: O(n²)
Space: O(n) - creates new list
-}
module Algorithms.InsertionSort
  ( sort
  , sortBy
  ) where

-- | Sort a list using insertion sort
sort :: Ord a => [a] -> [a]
sort = foldr insertSorted []

-- | Sort with custom comparison function
sortBy :: (a -> a -> Ordering) -> [a] -> [a]
sortBy cmp = foldr (insertSortedBy cmp) []

-- | Insert element into sorted list
insertSorted :: Ord a => a -> [a] -> [a]
insertSorted x [] = [x]
insertSorted x (y:ys)
  | x <= y    = x : y : ys
  | otherwise = y : insertSorted x ys

-- | Insert element with custom comparison
insertSortedBy :: (a -> a -> Ordering) -> a -> [a] -> [a]
insertSortedBy _ x [] = [x]
insertSortedBy cmp x (y:ys)
  | cmp x y /= GT = x : y : ys
  | otherwise     = y : insertSortedBy cmp x ys
